File: src/api/cloudKeyring.js
import { invoke } from '@tauri-apps/api/core';

export async function cloudKeyringSet(account, secret) {
    return await invoke('cloud_keyring_set', { account, secret });
}

export async function cloudKeyringGet(account) {
    return await invoke('cloud_keyring_get', { account });
}

export async function cloudKeyringDelete(account) {
    return await invoke('cloud_keyring_delete', { account });
}

File: src-tauri/src/cloud_keyring.rs
use keyring::Entry;

const SERVICE: &str = "cc.altron.mark2.cloud";

fn entry(account: &str) -> Result<Entry, String> {
    Entry::new(SERVICE, account).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn cloud_keyring_set(account: String, secret: String) -> Result<(), String> {
    entry(&account)?
        .set_password(&secret)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn cloud_keyring_get(account: String) -> Result<Option<String>, String> {
    match entry(&account)?.get_password() {
        Ok(s) => Ok(Some(s)),
        Err(keyring::Error::NoEntry) => Ok(None),
        Err(e) => Err(e.to_string()),
    }
}

#[tauri::command]
pub fn cloud_keyring_delete(account: String) -> Result<(), String> {
    match entry(&account)?.delete_credential() {
        Ok(()) => Ok(()),
        Err(keyring::Error::NoEntry) => Ok(()),
        Err(e) => Err(e.to_string()),
    }
}
